### ISSUE TITLE
Increase the PersistentVolume size for a Postgres Crunchydata cluster

### ISSUE THREAD

**Issue by paulbarbu** [CrunchyData/postgres-operator#1311]
Hello,

This question is similar to: https://github.com/CrunchyData/postgres-operator/issues/273
However I did not use the "manual installation method" and I;m not sure how to use the pgo.yaml file.

Using CrunchyData PGO on a bare metal cluster with Rook & Ceph for storage:
```
$ pgo version
pgo client version 4.2.0
pgo-apiserver version 4.2.0
```
Installed using ansible.

When I installed & deployed the Crunchydata Postgres Operator I forgot to modify the size of my storage (rook) in the ansible `inventory` file:
```storage9_name='rook'
storage9_access_mode='ReadWriteOnce'
storage9_size='1Gi'
storage9_type='dynamic'
storage9_class='rook-ceph-block'
storage9_fs_group=26
```
```
$ kubectl -n pgo-galaxy get pv
NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM                             STORAGECLASS      REASON   AGE
pvc-0cc55f2b-0ae3-4848-b4dc-bff7ce3a0d2b   1Gi        RWO            Delete           Bound    pgo-galaxy/pgocluster             rook-ceph-block            13d
pvc-8dd2de7c-7bee-459c-913a-7ee40c9bb328   1Gi        RWO            Delete           Bound    pgo-galaxy/pgocluster-pgbr-repo   rook-ceph-block            13d
```

Now I need more than 1GB of storage and need to resize the volumes.
How do I proceed?
Can I just edit the inventory file and the storage will be resized accordingly?
Should I sync the pgo.yaml file (located in `conf/postgres-operator`) with my ansible inventory and then run `kubectl apply -f pgo.yaml`?

I need to keep the 1GB of data I currently have.

**Comment by jkatz** [comment]
Hi @paulbarbu,

You will be able to do this once #1310 is committed. You can clone the PostgreSQL cluster into a new one with a larger PVC.

We will also introduce other ways of handling PVC resizing in future releases.

Thanks!

**Comment by paulbarbu** [comment]
> Hi @paulbarbu,
> [...]
> You can clone the PostgreSQL cluster into a new one with a larger PVC.
> [...]
> Thanks!
Could you please details this a bit?

**Comment by jkatz** [comment]
Hi @paulbarbu,

Please see this example:

------
### Clone a PostgreSQL Cluster to Different PVC Size

You can have a cloned PostgreSQL cluster use a different PVC size, which is
useful when moving your PostgreSQL cluster to a larger PVC. For example, to
clone a PostgreSQL cluster to a 256GB PVC, you can execute the following
command:

```shell
pgo clone hacluster newhacluster --pvc-size=256Gi
```

You can also have the cloned PostgreSQL cluster use a larger pgBackRest
backup repository by setting its PVC size. For example, to have a cloned
PostgreSQL cluster use a 1TB pgBackRest repository, you can execute the
following command:

```shell
pgo clone hacluster newhacluster --pgbackrest-pvc-size=1Ti
```
----

**Comment by paulbarbu** [comment]
Ok, so you were referring to the fact the I will be able to clone a cluster into another one while at the same time changing the PVC size in a future version?

Is there no way I could do it now?

My "pgo clone"  doesn't have those "size" arguments.

**Comment by rej-jsa** [comment]
I have a similar/related question to this one. On top of changing PVC size, is it (or will it be) possible (either via cloning, or in-place in a future version) to change a cluster's storage class? 

e.g. Migrating a cluster from a SATA-SSD backed Ceph pool to an NVMe backed Ceph pool.

**Comment by jkatz** [comment]
@paulbarbu The `pgo clone` feature I described was committed the other day and will be available in 4.3

PVC resizing itself is still relatively new to Kubernetes. The initial resizing work [went beta 1.11](https://kubernetes.io/blog/2018/07/12/resizing-persistent-volumes-using-kubernetes/) and was only made available for a group of volume plugins, and as such is not universal. There are also some settings that need to be enabled by a cluster administrator, in addition to the various spec configurations.

There is the added catch that for this to work, your Pod must go down, which as we know in PostgreSQL world, means downtime 😢 Now, in PostgreSQL Operator 4.2 and above, we can minimize that by performing a resize on the replicas before the primary, so as such we can failover with little to no disruption, but the risk for an outage is nonzero.

Online PVC expansion went alpha in 1.11 and [beta in 1.15](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#resizing-an-in-use-persistentvolumeclaim) - so we can at least be pretty confident that 1.15 clusters have this feature enabled. However, we would need to test for what it means with a PostgreSQL cluster -- as I read it, there is a potential downtime scenario as well. Additionally, we still need to support PostgreSQL clusters running on Kube 1.11 (OCP 3.11) for the time being, and of course would need to heavily test this :-)

This leads to one more solution, one that will not appear in Operator 4.3, but down the road, which is the Operator managing the PVC expansion itself. Effectively, this would be `pgo update cluster --pvc-size=blah`: under the surface, the Operator would spin up a temporary "replica" on a PVC of size "blah"; if the replica is successfully provisioned, we would then promote that replica to be the new primary. remove the old primary, and then add new replicas that have PVC of size `blah`. That would be agnostic to any of the Kubernetes methods, with the tradeoff that you need to have enough disk space available to add in one more PostgreSQL cluster temporarily.

In the interim, the `pgo clone --pvc-size` is a way that allows you to do this -- it's not as clean of an approach as it could be, but it's a start :wink: I also want you to know that this is a topic I have thought a lot about, and a lot of the work we have been doing over the past several months in the Operator is to get to this feature.

And if you need to resize now, my recommendation would be to create a replica of a primary (`pgo scale`), then modify the PVC spec of that specific new replica to resize its PVC. After such, manually fail over to that replica, and you have a larger cluster :smile:

**Comment by jkatz** [comment]
@rej-jsa That is a good idea, I have made a note of it. Thanks!

**Comment by paulbarbu** [comment]
@jkatz Thank you for the detailed response. That makes sense. I will try to resize using the `pgo scale` way if 4.3 doesn't get shipped soon.

**Comment by paulbarbu** [comment]
@jkatz Also, as a suggestion, I think that it may be worthwhile compiling all these scenarios into a FAQ or something similar.

While the work being done here is awesome, when such a case is encountered it quickly becomes a pain. At least having a section in the docs listing such cases and possible ways to go about them, would make the project much more usable and "explorable".

**Comment by jkatz** [comment]
That is a great suggestion. As this is an open source project, we also welcome patches too ;-)

**Comment by mesgar** [comment]
We've a similar but more complicated situation.
We've deployed 3 clusters on AKS and 
what we want to do is to create an alerting system that tells us if the disk usages is above 85% add nGB (e.g. 100GB) disk space  automatically.
Our aim is zero human interference. 
![image](https://user-images.githubusercontent.com/10726024/107782321-424b1d00-6d8c-11eb-871d-5db63b62a1fe.png)
We're thinking about a batch system to df the PVC size and resize it when needed.
pgo df
and then
pgo sacle or pgo clone?
BTW, pgo df returns Error!?
![image](https://user-images.githubusercontent.com/10726024/107786081-bdaecd80-6d90-11eb-91de-77a03346e4b8.png)

![image](https://user-images.githubusercontent.com/10726024/107785320-c652d400-6d8f-11eb-9a17-91ed926a1c69.png)

What is the least risky way to do this?

**Comment by JanCizmar** [comment]
Hello, why not to resize the PVC following this [Resizing Persistent Volumes using Kubernetes](https://kubernetes.io/blog/2018/07/12/resizing-persistent-volumes-using-kubernetes/)  ? It seems working to me. Is there any drawback?

**Comment by mesgar** [comment]
Thanks @JanCizmar 
This is how I created my cluster.
pgo create cluster iot-db \
  --service-type=LoadBalancer \
  --replica-count 3 \
  --pvc-size 100.10Gi
and I've "allowVolumeExpansion: true" by default.
![image](https://user-images.githubusercontent.com/10726024/107838806-fd0a0800-6dea-11eb-9e57-c50468b2de65.png)

before this I had a 1Gb PVC (which is the default size) and when I tried to Import a large database I got an error saying there no space left on the disk.
I edited my cluster yaml file (kubectl -n pgo edit pvc iot-db) and changed the storage size but nothing happened.
I tried to scale down/up and update cluster which didn't work either.
I restarted nodes (vmss) and it didn't work too.
thats why I had to recreate my cluster and passing through [--pvc-size 100.10Gi] parameter when creating my cluster.
I changed the current size to 200GB and did everything explained above.
but when i [kubectl -n pgo get pvc iot-db -o yaml]
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 200Gi
  storageClassName: default
  volumeMode: Filesystem
  volumeName: pvc-b19b314c-e0ae-4a2d-9800-3ca8393b90e2
status:
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 100Gi
  conditions:
  - lastProbeTime: null
    lastTransitionTime: 2021-02-13T02:17:24Z
    status: "True"
    type: Resizing
  phase: Bound

It shows resizing. 
do I need to delete my pods manually?
Our aim is zero downtime and zero human interference.
